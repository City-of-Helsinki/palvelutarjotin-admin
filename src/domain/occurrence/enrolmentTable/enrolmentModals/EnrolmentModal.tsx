import classNames from 'classnames';
import { Button, Checkbox, IconCross, Notification, TextArea } from 'hds-react';
import React, { ReactNode } from 'react';
import { useTranslation } from 'react-i18next';
import ReactModal from 'react-modal';

import EnrolleesList from './EnrolleesList';
import styles from './enrolmentModals.module.scss';

export type EnrolleeProps = {
  personName?: string;
  studyGroupName?: string | null;
  studyLevel?: string | null;
  groupSize?: number | null;
  amountOfAdult?: number | null;
};

interface EnrolmentModalProps {
  onClose: () => void;
  title: string;
  appElement?: HTMLElement;
  enrollees?: EnrolleeProps[];
  noteText?: string | ReactNode;
  noteType?: 'success' | 'decline';
  noteSection?: boolean;
  submitButtonText: string;
  handleSubmit: () => void;
  preview?: ReactNode | null;
  handlePreviewClick?: () => void;
  onMessageTextChange?: (message: string) => void;
  showPreviewButton?: boolean;
  messageText?: string;
  submitting?: boolean;
}

const EnrolmentModal: React.FC<EnrolmentModalProps> = ({
  children,
  title,
  appElement,
  enrollees,
  noteText,
  noteType = 'success',
  noteSection,
  submitButtonText,
  preview,
  messageText,
  showPreviewButton,
  submitting = false,
  handleSubmit,
  handlePreviewClick,
  onMessageTextChange,
  onClose,
}) => {
  const { t } = useTranslation();
  const [showMessageTextArea, setShowMessageTextArea] = React.useState(false);
  const hasEnrollees = enrollees && enrollees.length > 0;

  const renderModalContent = () => {
    return (
      <>
        {noteText && (
          <div
            className={classNames({
              [styles.infoNoteSuccess]: noteType === 'success',
              [styles.infoNoteDecline]: noteType === 'decline',
            })}
          >
            {noteText}
          </div>
        )}
        {hasEnrollees && <EnrolleesList enrollees={enrollees} />}
        {noteSection && (
          <div className={styles.addNoteSection}>
            <Checkbox
              className={styles.noteCheckbox}
              onChange={() => setShowMessageTextArea(!showMessageTextArea)}
              checked={showMessageTextArea}
              id="add-note"
              label={t('enrolment.enrolmentModal.addMessage')}
            />
            {noteType === 'success' && (
              <div className={styles.formRow}>
                <Notification
                  label={t('eventOccurrenceForm.infoTitleAutoAcceptance')}
                >
                  {t('eventOccurrenceForm.infoContentAutoAcceptance')}
                </Notification>
              </div>
            )}
            {showMessageTextArea && (
              <div>
                <TextArea
                  className={styles.noteTextArea}
                  value={messageText}
                  id="note-text-area"
                  label={t('enrolment.enrolmentModal.messageToParticipants')}
                  onChange={(e) => onMessageTextChange?.(e.target.value)}
                />
              </div>
            )}
          </div>
        )}
        <div className={styles.buttonsContainer}>
          <Button variant="secondary" onClick={onClose}>
            {t('enrolment.enrolmentModal.cancelEnrolment')}
          </Button>
          <div className={styles.buttonsRight}>
            {/* TODO: preview functionality */}
            {showPreviewButton && (
              <Button
                iconLeft={<div></div>}
                variant="supplementary"
                onClick={handlePreviewClick}
              >
                {t('enrolment.enrolmentModal.preview')}
              </Button>
            )}
            <Button
              variant="primary"
              onClick={handleSubmit}
              disabled={submitting}
            >
              {submitButtonText}
            </Button>
          </div>
        </div>
      </>
    );
  };

  return (
    <ReactModal
      isOpen
      bodyOpenClassName={styles.bodyOpen}
      className={classNames(styles.modal)}
      overlayClassName={styles.overlay}
      appElement={appElement}
    >
      <div className={styles.titleRow}>
        <p>{title}</p>
        <button className={styles.closeButton} onClick={onClose}>
          <IconCross />
          {t('common.buttonClose')}
        </button>
      </div>
      {preview ? preview : renderModalContent()}
      {children}
    </ReactModal>
  );
};

export default EnrolmentModal;
