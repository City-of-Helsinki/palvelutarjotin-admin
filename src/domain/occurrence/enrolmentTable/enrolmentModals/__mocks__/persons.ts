export default [
  {
    id: 'UGVyc29uTm9kZTpkNTUyMWIwOC1iZDJhLTQ3MTctOGM1NS1mZThhYzJhY2YzYmE=',
    emailAddress: 'maili@hotmail.com',
    name: 'Nimi Niminen',
    phoneNumber: '3457675654',
    language: 'FI',
    __typename: 'PersonNode',
  },
  {
    id: 'UGVyc29uTm9TpkNTUyMWIwOC1iZDJhLTQ3MTctOGM1NS1mZThhYzJhY2YzYmE=',
    emailAddress: 'maili2@hotmail.com',
    name: 'Toka Tokanen',
    phoneNumber: '876436675',
    language: 'FI',
    __typename: 'PersonNode',
  },
  {
    id: 'UGVyc29uTm9kZTpkNTUyMWIwOC1iZDJhLTQ3ctOGM1NS1mZThhYzJhY2YzYmE=',
    emailAddress: 'maili3@hotmail.com',
    name: 'Kolme Kolmonen',
    phoneNumber: '123234345',
    language: 'FI',
    __typename: 'PersonNode',
  },
];
