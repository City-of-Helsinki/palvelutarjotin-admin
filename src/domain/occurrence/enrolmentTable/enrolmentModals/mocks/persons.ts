const persons = [
  {
    personName: 'Nimi Niminen',
  },
  {
    personName: 'Toka Tokanen',
  },
  {
    personName: 'Kolme Kolmonen',
  },
];

export default persons;
