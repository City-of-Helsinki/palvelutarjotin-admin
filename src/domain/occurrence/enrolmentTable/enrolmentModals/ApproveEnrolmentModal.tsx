import { Button, Checkbox, TextArea } from 'hds-react';
import React from 'react';
import { useTranslation } from 'react-i18next';

import { PersonFieldsFragment } from '../../../../generated/graphql';
import EnrolleesList from './EnrolleesList';
import EnrolmentModal from './EnrolmentModal';
import styles from './enrolmentModals.module.scss';

interface ApproveModalProps {
  isOpen: boolean;
  onClose: () => void;
  approveEnrolment: () => void;
  enrollees?: PersonFieldsFragment[];
}

const ApproveEnrolmentModal: React.FC<ApproveModalProps> = ({
  isOpen,
  onClose,
  approveEnrolment,
  enrollees,
}) => {
  const { t } = useTranslation();
  const [messageText, setMessageText] = React.useState('');
  const [showMessageTextArea, setShowMessageTextArea] = React.useState(false);

  const handlePreview = () => {
    alert('TODO: handle preview');
  };

  return (
    <EnrolmentModal
      isOpen={isOpen}
      onClose={onClose}
      title={t('enrolment.enrolmentModal.approveEnrolment')}
    >
      <div className={styles.infoNoteSuccess}>
        {t('enrolment.enrolmentModal.approveEnrolmentNote')}
      </div>
      <EnrolleesList enrollees={enrollees} />
      <div className={styles.addNoteSection}>
        <Checkbox
          className={styles.noteCheckbox}
          onChange={() => setShowMessageTextArea(!showMessageTextArea)}
          checked={showMessageTextArea}
          id="add-note"
          label={t('enrolment.enrolmentModal.addMessage')}
        />
        {showMessageTextArea && (
          <TextArea
            className={styles.noteTextArea}
            value={messageText}
            id="note-text-area"
            label={t('enrolment.enrolmentModal.messageToParticipants')}
            onChange={(e) => setMessageText(e.target.value)}
          />
        )}
      </div>
      <div className={styles.buttonsContainer}>
        <Button variant="secondary" onClick={onClose}>
          {t('enrolment.enrolmentModal.cancelEnrolment')}
        </Button>
        <div className={styles.buttonsRight}>
          {/* TODO: preview functionality */}
          <Button variant="supplementary" onClick={handlePreview}>
            {t('enrolment.enrolmentModal.preview')}
          </Button>
          <Button variant="primary" onClick={approveEnrolment}>
            {t('enrolment.enrolmentModal.sendConfirmationMessage')}
          </Button>
        </div>
      </div>
    </EnrolmentModal>
  );
};

export default ApproveEnrolmentModal;
