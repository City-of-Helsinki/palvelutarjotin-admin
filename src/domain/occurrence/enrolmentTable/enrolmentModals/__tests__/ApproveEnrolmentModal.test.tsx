import userEvent from '@testing-library/user-event';
import * as React from 'react';

import messages from '../../../../../domain/app/i18n/fi.json';
import {
  actWait,
  render,
  screen,
  waitFor,
} from '../../../../../utils/testUtils';
import ApproveEnrolmentModal from '../ApproveEnrolmentModal';
import { EnrolleeProps } from '../EnrolmentModal';
import persons from '../mocks/persons';

it('matches snapshot', async () => {
  const { baseElement } = render(
    <ApproveEnrolmentModal
      enrolmentId="123"
      onClose={jest.fn()}
      approveEnrolment={jest.fn()}
      appElement={document.body}
    />
  );

  await actWait();

  await waitFor(() => {
    // A small "hack" to wait for modal to open fully
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      baseElement.querySelector('.ReactModal__Content--after-open')
    ).toBeInTheDocument();
  });

  expect(baseElement).toMatchSnapshot();
});

it('renders correctly and calls approve enrolment handler', async () => {
  const onCloseHandler = jest.fn();
  const approveEnrolmentHandler = jest.fn();
  render(
    <ApproveEnrolmentModal
      enrolmentId="123"
      onClose={onCloseHandler}
      approveEnrolment={approveEnrolmentHandler}
      appElement={document.body}
    />
  );

  await actWait();

  expect(
    screen.getByText(messages.enrolment.enrolmentModal.approveEnrolment)
  ).toBeInTheDocument();

  expect(
    screen.getByText(messages.enrolment.enrolmentModal.approveEnrolmentNote)
  ).toBeInTheDocument();

  expect(
    screen.getByText(
      /vahvistusviesti sisältää automaattisesti seuraavat tiedot/i
    )
  ).toBeInTheDocument();
  expect(
    screen.getByText(
      /personoitu tervehdys, ilmoittautuminen vahvistettu, tapahtuman tiedot, aika, varattujen paikkojen lukumäärä, kieli, paikka, osoite, järjestäjän yhteystiedot\./i
    )
  ).toBeInTheDocument();

  const approveEnrolmentButton = screen.getByText(
    messages.enrolment.enrolmentModal.sendConfirmationMessage
  );
  await userEvent.click(approveEnrolmentButton);

  expect(approveEnrolmentHandler).toHaveBeenCalledTimes(1);
});

it('opens message section when checkbox is clicked and text can be written', async () => {
  render(
    <ApproveEnrolmentModal
      enrolmentId="123"
      onClose={jest.fn()}
      approveEnrolment={jest.fn()}
      appElement={document.body}
    />
  );

  await actWait();

  const addMessageCheckbox = screen.getByLabelText(/lisää viesti/i);
  await userEvent.click(addMessageCheckbox);

  const messageTextArea = screen.getByLabelText(
    messages.enrolment.enrolmentModal.messageToParticipants
  );

  expect(messageTextArea).toBeInTheDocument();

  await userEvent.type(messageTextArea, 'Tässä testiviesti');

  expect(messageTextArea).toHaveTextContent('Tässä testiviesti');
});

it('renders enrollees list correctly', async () => {
  render(
    <ApproveEnrolmentModal
      enrolmentId="123"
      onClose={jest.fn()}
      approveEnrolment={jest.fn()}
      appElement={document.body}
      enrollees={persons as EnrolleeProps[]}
    />
  );

  await actWait();

  persons.forEach((person) => {
    expect(screen.getByText(person.personName)).toBeInTheDocument();
  });
});
