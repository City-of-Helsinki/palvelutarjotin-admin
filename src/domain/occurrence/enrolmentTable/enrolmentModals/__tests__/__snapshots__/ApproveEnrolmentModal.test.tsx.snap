// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<body
  aria-hidden="true"
  class="bodyOpen"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay ReactModal__Overlay--after-open overlay"
    >
      <div
        aria-modal="true"
        class="ReactModal__Content ReactModal__Content--after-open modal"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="titleRow"
        >
          <p>
            Hyväksy ilmoittautuminen
          </p>
          <button
            class="closeButton"
          >
            <svg
              class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik"
              role="img"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fill-rule="evenodd"
              >
                <rect
                  height="24"
                  width="24"
                />
                <polygon
                  fill="currentColor"
                  points="18 7.5 13.5 12 18 16.5 16.5 18 12 13.5 7.5 18 6 16.5 10.5 12 6 7.5 7.5 6 12 10.5 16.5 6"
                />
              </g>
            </svg>
            Sulje
          </button>
        </div>
        <div
          class="infoNoteSuccess"
        >
          Valittujien ilmoittautujien osallistuminen vahvistetaan ja heille lähetetään vahvistusviesti tapahtuman tiedoilla.
        </div>
        <div
          class="addNoteSection"
        >
          <div
            class="Checkbox-module_checkbox__3L0GR checkbox_hds-checkbox__9HMCz noteCheckbox"
          >
            <input
              class="Checkbox-module_input__3VZvy checkbox_hds-checkbox__input__1w0pu"
              id="add-note"
              type="checkbox"
              value=""
            />
            <label
              class="Checkbox-module_label__L5AN1 checkbox_hds-checkbox__label__3HoD3"
              for="add-note"
            >
              Lisää viesti
            </label>
          </div>
          <div
            class="formRow"
          >
            <section
              aria-atomic="true"
              aria-label="Notification"
              class="Notification-module_notification__2hN3H notification_hds-notification__2DQPN"
            >
              <div
                class="Notification-module_content__ZZKvE notification_hds-notification__content__5ylSS"
              >
                <div
                  aria-level="2"
                  class="Notification-module_label__1bhNu notification_hds-notification__label__2t1lR"
                  role="heading"
                >
                  <svg
                    aria-hidden="true"
                    class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Notification-module_icon__1ans7 notification_hds-icon__29HQM"
                    role="img"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      fill="none"
                      fill-rule="evenodd"
                    >
                      <rect
                        height="24"
                        width="24"
                      />
                      <path
                        d="M12,3 C7.02943725,3 3,7.02943725 3,12 C3,16.9705627 7.02943725,21 12,21 C16.9705627,21 21,16.9705627 21,12 C21,7.02943725 16.9705627,3 12,3 Z M13,10 L13,16.5 L15,16.5 L15,18 L9,18 L9,16.5 L11,16.5 L11,11.5 L9,11.5 L9,10 L13,10 Z M11.8125239,6 C12.5373649,6 13.125,6.58759799 13.125,7.3125 C13.125,8.03735415 12.5373649,8.625 11.8125239,8.625 C11.0876351,8.625 10.5,8.03735415 10.5,7.3125 C10.5,6.58759799 11.0876351,6 11.8125239,6 Z"
                        fill="currentColor"
                      />
                    </g>
                  </svg>
                  Vahvistusviesti sisältää automaattisesti seuraavat tiedot
                </div>
                <div
                  class="Notification-module_body__9BSAD notification_hds-notification__body__3cobH"
                >
                  Personoitu tervehdys, ilmoittautuminen vahvistettu, tapahtuman tiedot, aika, varattujen paikkojen lukumäärä, kieli, paikka, osoite, järjestäjän yhteystiedot.
                </div>
              </div>
            </section>
          </div>
        </div>
        <div
          class="buttonsContainer"
        >
          <button
            class="Button-module_button__1msFE button_hds-button__2A0je Button-module_secondary__1nABp button_hds-button--secondary__1NOWS"
            type="button"
          >
            <span
              class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
            >
              Peruuta
            </span>
          </button>
          <div
            class="buttonsRight"
          >
            <button
              class="Button-module_button__1msFE button_hds-button__2A0je Button-module_supplementary__3YKiS button_hds-button--supplementary__GcHcV"
              type="button"
            >
              <div
                aria-hidden="true"
                class="Button-module_icon__O-h7R button_hds-icon__17j8Z"
              >
                <div />
              </div>
              <span
                class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
              >
                Esikatselu
              </span>
            </button>
            <button
              class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO"
              type="button"
            >
              <span
                class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
              >
                Lähetä vahvistusviesti
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
